fail_over: protect against a segmentation fault

sumit-bose opened the pull request:
A missing server name in struct fo_server will cause a segmentation
fault. Currently it is unclear why the server name is missing at this
point. To avoid the segmentation fault it is checked before if the
server name is missing. Additionally the state of some internal
structures is added to the debug logs to help debugging why the server
name is missing.

Resolves: https://github.com/SSSD/sssd/issues/6659

aplopez commented on src/providers/data_provider_fo.c:
```diff
@@ -675,6 +683,12 @@ errno_t be_resolve_server_process(struct tevent_req *subreq,
+    } else {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing server name.\n");
+        dump_be_svc_data(state->svc);
+        dump_fo_server(state->srv);
```
Won't `dump_fo_server_list()` also log the server `state->srv`? Why also logging it separately here?

sumit-bose replied:
Hi,

with just dumping the list it would not be clear which list entry corresponds to `state->srv`. Additionally, since something unexpected it going on, `state->srv` might still have back-links to list entries but is not on the list anymore, i.e. no other list items will have links to `state->srv`.

bye,
Sumit

aplopez commented on src/providers/fail_over.c:
```diff
@@ -200,6 +200,31 @@ str_srv_data_status(enum srv_lookup_status status)
+                             "port_status [%s] common->name [%s].\n",
+                             srv->primary ? "true" : "false", srv->port,
+                             str_port_status(srv->port_status),
+                             srv->common == NULL ? "common is NULL"
```
When `srv->common` is `NULL`, the message will be `fo_server: ... common->name [common is NULL].` Wouldn't it better be just `common->name [NULL]`?

sumit-bose replied:
`common->name [NULL]` would indicate that `srv->common->name` is `NULL` and `common->name [common is NULL]` means that already `srv->common` is `NULL`.

alexey-tikhonov replied:
> common->name [NULL] would indicate that srv->common->name is NULL

While most compilers will handle it well, I think strictly speaking providing NULL 'srv->common->name' to `%s` isn't "legal".

alexey-tikhonov replied:
Would it make sense to print everything from 'srv_data' and 'common' fields?

alexey-tikhonov commented on src/providers/data_provider_fo.c:
```diff
@@ -594,6 +594,14 @@ static void be_resolve_server_done(struct tevent_req *subreq)
     tevent_req_error(req, ret);
 }
 
+void dump_be_svc_data(struct be_svc_data *svc)
```
This could be `static`?

alexey-tikhonov replied:
In all helper functions argument could be a pointer to const struct...

sumit-bose replied:
done

alexey-tikhonov commented on src/providers/data_provider_fo.c:
```diff
@@ -594,6 +594,14 @@ static void be_resolve_server_done(struct tevent_req *subreq)
+    DEBUG(SSSDBG_OP_FAILURE, "be_svc_data: name=[%s] last_good_srv=[%s] "
+                             "last_good_port=[%d] last_status_change=[%lld]\n",
+                             svc->name, svc->last_good_srv, svc->last_good_port,
+                             (long long) svc->last_status_change);
```
`last_status_change` is `time_t`
Would it make sense to rebase (and use) on top of recently merged 54aabca0b65037600cf5238c452698e81a407bc5?

sumit-bose replied:
done

alexey-tikhonov commented on src/providers/data_provider_fo.c:
```diff
@@ -675,6 +683,12 @@ errno_t be_resolve_server_process(struct tevent_req *subreq,
               fo_get_server_str_name(state->srv), ipaddr,
               srvaddr->addr_list[0]->ttl);
+    } else {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing server name.\n");
```
Idea is that this ^^ line triggers a backtrace and then following lines are printed with default log level cleanly (without backtrace)?

sumit-bose replied:
Hi,

yes, but mainly I wanted a message even if the default debug level was lowered.

bye,
Sumit

alexey-tikhonov commented:
Hi @sumit-bose,
thank you for the updates.

There is one unaddressed comment (https://github.com/SSSD/sssd/pull/6660/files#r1171762657) left.

sumit-bose commented:
> Hi @sumit-bose, thank you for the updates.
> 
> There is one unaddressed comment (https://github.com/SSSD/sssd/pull/6660/files#r1171762657) left.

Ah, sorry, I missed this. I dumped the content of struct srv_data as well. common contains basically pointers where I think that they might not help much when just printed without additional context. Also `srv->common->name == NULL` is handled.

bye,
Sumit

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/6660

* `master`
    * 8a8869994745429b3f5535a5d0b91f1d0b2fa723 - fail_over: protect against a segmentation fault
* `sssd-2-9`
    * f63a54c3d89866949810d5bcdb09bf2edd3ced02 - fail_over: protect against a segmentation fault
